pkg/wasm/cache.go

```diff
@@ -100,52 +102,69 @@ func NewLocalFileCache(dir string, purgeInterval, moduleExpiry time.Duration) *L
 
 // Get returns path the local Wasm module file.
 func (c *LocalFileCache) Get(downloadURL, checksum string, timeout time.Duration) (string, error) {
-	url, err := url.Parse(downloadURL)
-	if err != nil {
-		return "", fmt.Errorf("fail to parse Wasm module fetch url: %s", downloadURL)
-	}
 	// Construct Wasm cache key with downloading URL and provided checksum of the module.
 	key := cacheKey{
 		downloadURL: downloadURL,
 		checksum:    checksum,
 	}
 
-	switch url.Scheme {
-	case "http", "https":
-		// First check if the cache entry is already downloaded.
-		if modulePath := c.getEntry(key); modulePath != "" {
-			return modulePath, nil
-		}
+	// First check if the cache entry is already downloaded.
+	if modulePath := c.getEntry(key); modulePath != "" {
+		return modulePath, nil
+	}
+
+	// If not, fetch images.
+	u, err := url.Parse(downloadURL)
+	if err != nil {
+		return "", fmt.Errorf("fail to parse Wasm module fetch url: %s", downloadURL)
+	}
 
-		// If the module is not available locally, download the Wasm module with http fetcher.
-		b, err := c.httpFetcher.Fetch(downloadURL, timeout)
+	// Byte array of Wasm binary.
+	var b []byte
+	// Hex-Encoded sha256 checksum of binary.
+	var dChecksum string
+	switch u.Scheme {
+	case "http", "https":
+		// Download the Wasm module with http fetcher.
+		b, err = c.httpFetcher.Fetch(downloadURL, timeout)
 		if err != nil {
 			wasmRemoteFetchCount.With(resultTag.Value(downloadFailure)).Increment()
 			return "", err
 		}
 
 		// Get sha256 checksum and check if it is the same as provided one.
-		dChecksum := fmt.Sprintf("%x", sha256.Sum256(b))
+		sha := sha256.Sum256(b)
+		dChecksum = hex.EncodeToString(sha[:])
 		if checksum != "" && dChecksum != checksum {
 			wasmRemoteFetchCount.With(resultTag.Value(checksumMismatch)).Increment()
 			return "", fmt.Errorf("module downloaded from %v has checksum %v, which does not match: %v", downloadURL, dChecksum, checksum)
 		}
+	case "oci":
+		ctx, cancel := context.WithTimeout(context.Background(), timeout)
+		defer cancel()
+		// TODO: support imagePullSecret and pass it to ImageFetcherOption.
+		fetcher := NewImageFetcher(ctx, ImageFetcherOption{})
+		b, err = fetcher.Fetch(u.Host+u.Path, checksum)
+		if err != nil {
+			return "", fmt.Errorf("could not fetch Wasm OCI image: %v", err)
```

Comment:
Can you increment `wasmRemoteFetchCount` counter here with download failure?

Reply:
done https://github.com/istio/istio/pull/34430/commits/0d5f0bd97629bc8414eb3d8a967b511d8d3336b1 thanks!